File: public/assets/js/votes.js
function splitNumber(number) {

    // Convertir el número a su representación binaria de 8 bits
    let binary = number.toString(2).padStart(8, '0');

    // Obtener los primeros 4 bits y los últimos 4 bits
    let first4Bits = binary.slice(1, 4);
    let last4Bits = binary.slice(4, 8);

    // Convertir los primeros 4 bits y los últimos 4 bits de binario a decimal
    let first4Decimal = parseInt(first4Bits, 2);
    let last4Decimal = parseInt(last4Bits, 2);
    votes = ['A', 'B', 'C', 'D', 'E', 'F']
    voteString = votes[first4Decimal]
    return {
        vote: voteString,
        number: last4Decimal
    };
}
function handleReport(data, id) {


    let result = splitNumber(data[1]);

    control = (data[0] - 32) * 16 + result.number
    return {
        control: control,
        vote: result.vote
    }

}

document.getElementById('connectBtn').addEventListener('click', async () => {
    try {

        // Solicitar el acceso a un dispositivo HID
        const devices = await navigator.hid.requestDevice({ filters: [] });

        if (devices.length === 0) {
            console.log('No device selected');
            return;
        }

        // Conectar al dispositivo seleccionado
        const device = devices[0];
        await device.open();
        console.log(`Connected to ${device.productName}`);

        // Mostrar información del dispositivo
        const deviceInfo = document.getElementById('deviceInfo');


        // Escuchar los datos del dispositivo
        const dataStream = document.getElementById('dataStream');
        flag = 4;
        let ignoreSecond = false;
        let awaitSecond = false;
        let head = 0
        device.oninputreport = (event) => {
            flag++;

            if (flag > 4) {

                if (ignoreSecond) {
                    ignoreSecond = false;
                    return;
                }

                // Extraer el buffer de datos y el reportId del evento
                const { data, reportId } = event;
                console.log('reportId', reportId);
                const dataArray = new Uint8Array(data.buffer);

                // Obtener los primeros 3 elementos
                let receivedData = dataArray.slice(0, 3);
                // hay tres casos de respuesta
                //1. caso con reportID=3, viene un unico mensaje con toda la informacion correcta
                //2. caso con reportId=2, vienen dos mensajes el primero tiene reportId=2 y el otro en 1, solo el primero tiene la informacion correcta
                //3. caso con reportId=1, vienen dos, ambos tienen la informacion parcial, se tiene en cuenta el
                //primer elementol de cada uno

                //solo se maneja si no es 3
                if (reportId != 3) {
                    //si es uno se guarda el primer elemento y pasa
                    if (reportId == 1) {
                        awaitSecond = true;
                        head = receivedData[0]
                        return;
                    }
                    //si es dos ordena ignorar el siguiente
                    else if (reportId == 2 && !awaitSecond) {
                        ignoreSecond = true;
                    }
                    //es el segundo mensaje del caso 3
                    else if (reportId == 2 && awaitSecond ) {
                        let dataPass = [head, receivedData[0]]
                        receivedData = dataPass;
                        awaitSecond = false;
                    }
                }
                let dataResult = handleReport(receivedData, reportId)
                if(['A','B','C','D','E','F'].includes(dataResult.vote)){
                    dataStream.value += `\n${dataResult.control}-${dataResult.vote}`;
                }


            }
        };



        // Enviar comandos al dispositivo
        document.getElementById('sendCommandBtn').addEventListener('click', async () => {
            flag = 0;
            if ($controles = 100) {
                first = [0x03, 0x74, 0x86, 0xf2]
            }

            const commands = [

                [0x03, 0x55, 0x80, 0xd5],
                [0x0b, 0x5e, 0x80, 0xde, 0xb2, 0xb0, 0xb0, 0xb2, 0xb0, 0xb1, 0xb1, 0xb7],
                [0x06, 0x35, 0x96, 0x02, 0x10, 0x00, 0x10],
                [0x03, 0x03, 0x89, 0x8a],
                [0x03, 0x52, 0x86, 0xd4],
                [0x03, 0x5a, 0x80, 0xda]
            ];
            for (const command of commands) {
                try {


                    // Convertir el comando en un array de bytes
                    const commandArray = new Uint8Array(command);

                    // Enviar el comando al dispositivo
                    await device.sendReport(0x6, commandArray); // El primer argumento es el report ID, ajusta si es necesario
                    console.log('Command sent:', command);

                    // Esperar un poco antes de enviar el siguiente comando
                    await new Promise(resolve => setTimeout(resolve, 500));
                } catch (error) {
                    console.log('Error al escribir el comando: ', error)
                }

            }

        });

        // Enviar comandos al dispositivo
        document.getElementById('stopReadMode').addEventListener('click', async () => {

            const commands = [
                [0x03, 0x5b, 0x80, 0xdb],
                [0x03, 0x5b, 0x80, 0xdb],
                [0x03, 0x03, 0x82, 0x81]
            ];
            for (const command of commands) {
                try {
                    // Convertir el comando en un array de bytes
                    const commandArray = new Uint8Array(command);

                    // Enviar el comando al dispositivo
                    await device.sendReport(0x6, commandArray); // El primer argumento es el report ID, ajusta si es necesario
                    console.log('Command sent:', command);

                    // Esperar un poco antes de enviar el siguiente comando
                    await new Promise(resolve => setTimeout(resolve, 500));
                } catch (error) {
                    console.log('Error al escribir el comando: ', error)
                }

            }

        });

        document.getElementById('cleanBtn').addEventListener('click', async () => {
            dataStream.value = ``;
            flag = 4
        });


    } catch (error) {
        console.error('Error accessing HID device:', error);
    }
});

File: scripts/nova.python/votes.py

import hid
import time
import mysql.connector
import sys
import os

# Listar dispositivos HID conectados
# for device in hid.enumerate():
#     print(f"Device:{device['product_string']} Vendor ID: {device['vendor_id']}, Product ID: {device['product_id']}")

def connectDB():
    # Conectarse a la base de datos
    conn = mysql.connector.connect(
        host="localhost",
        user="root",
        password="",
        database="novaDB"
    )

    cursor = conn.cursor()



    return cursor,conn


def split_number(number):
    # Convertir el número a su representación binaria de 8 bits
    binary = format(number, '08b')

    # Obtener los primeros 4 bits y los últimos 4 bits
    first4_bits = binary[1:4]  # Aquí es del índice 0 al 3
    last4_bits = binary[4:]   # Aquí es del índice 4 al final

    # Convertir los primeros 4 bits y los últimos 4 bits de binario a decimal
    first4_decimal = int(first4_bits, 2)
    last4_decimal = int(last4_bits, 2)
    # Lista de votos
    votes = ['A', 'B', 'C', 'D', 'E', 'F']

    # Obtener el voto
    vote_string = votes[first4_decimal]

    return {
        'vote': vote_string,
        'number': last4_decimal
    }

def handleReport(data, id) :


    result = split_number(data[1])

    control = (data[0] - 32) * 16 + result['number']
    return {
        'control_id': control,
        'vote': result['vote']
    }

vid = 4292  # Reemplazar con tu vendor_id
pid = 6169  # Reemplazar con tu product_id
def connect():
    try:
        # Abrir conexión con el dispositivo
        device = hid.device()
        device.open(vid, pid)
        device.set_nonblocking(1)  # Configurar modo no bloqueante
        return device

    except OSError as e:
        print(f"Error al conectar con el dispositivo HID: {e}")
        return False

def sendComands(device):

    commands = [
                [0x03, 0x78, 0x8c, 0xf4],
                [0x03, 0x55, 0x80, 0xd5],
                [0x0b, 0x5e, 0x80, 0xde, 0xb2, 0xb0, 0xb0, 0xb2, 0xb0, 0xb1, 0xb1, 0xb7],
                [0x06, 0x35, 0x96, 0x02, 0x10, 0x00, 0x10],
                [0x03, 0x03, 0x89, 0x8a],
                [0x03, 0x52, 0x86, 0xd4],
                [0x03, 0x5a, 0x80, 0xda]]
    for command in commands:
        device.write(command)
        time.sleep(.5)

def sendComandsToClose(device):

    commands = [
                [0x03, 0x5b, 0x80, 0xdb],
                [0x03, 0x5b, 0x80, 0xdb],
                [0x03, 0x03, 0x82, 0x81]]
    for command in commands:
        device.write(command)
        time.sleep(.5)


def getArgs():
    # Verifica si hay suficientes argumentos
    if len(sys.argv) > 1:
        # Obtiene el primer argumento (titulo)
        controls = sys.argv[1]
        return controls
    else:
        print("1000")


def setProcessId():
    pid_file = 'script_pid.txt'

    # Escribe el PID en un archivo
    with open(pid_file, 'w') as f:
        f.write(str(os.getpid()))

def main():
    try:
        print('Estableciendo Conexión con el Dispositivo y la BD...')
        cursor,conn = connectDB()
        device=connect()
        if device:

            sendComands(device)
            time.sleep(.5)
            myFlag=0
            ignoreSecond = False
            awaitSecond = False
            head = 0
            i=0
            caso=0
            while True:

                data = device.read(64)  # Leer un bloque de datos de 64 bytes

                if data and myFlag:
                    print(data[:4])
                    reportId=data[0]
                    values=[data[1],data[2],data[3]]
                    #caso 2.1 se ignora el siguiente mensaje si su report id es 1
                    if ignoreSecond:
                        ignoreSecond = False
                        if reportId == 1 or (values[0]>80 and reportId == 1) :
                            continue
                    # hay multiples casos de respuesta
                    #siempre necesitamos dos valores uno que indica el grupo: va de 32 a 80
                    #otro que indica el control y el voto
                    #CASO 1, comienza con reportId 1 y el siguiente mensaje es 2 o 3, se toma el primer valor de cada mensaje
                    if reportId == 1 and values[0]<=80  :
                        caso=1
                        awaitSecond = True
                        head = values[0]
                        continue
                    #CASO 1.1, se lee el siguiente mensaje y se calcula el resultado con eso
                    elif (reportId == 2 or reportId == 3)  and awaitSecond :

                        values = [head, values[0]]
                        awaitSecond = False
                        head=0
                    #CASO 2, vienen dos mensajes, el primer con id 2 y el otro en 1, se toman los dos primeros valores del primer mensaje
                    elif reportId == 2 and values[0]<80 :
                        caso=2
                        ignoreSecond = True
                    #CASO 3, con report id 2 viene un unico mensaje, se toma el ultimo valor y el primero en ese orden
                    elif reportId == 2 and values[2]<80:
                        caso=3
                        dataPass=[values[2],values[0]]
                        values=dataPass
                    #CASO 4, con reportId 3 viene un mensaje se toman sus dos primeros valores
                    elif reportId == 3 and values[0]<=80:
                        caso=4
                    #CASO 5, dos mensajes con reportId 3 y 1 los dos ultimos valores deben ser iguales y esos se toman
                    elif reportId == 3 and values[1]<80:
                        caso=5
                        dataPass =[values[1],values[2]]
                        awaitSecond= True
                        continue
                    #CASO 5.1, se recibe el segundo valor y se compara que sean iguales
                    elif reportId == 1 and values[0]>80 and awaitSecond:

                        awaitSecond= False
                        if dataPass[0]==values[1] and dataPass[1]==values[2]:
                            values=dataPass
                        else:
                            continue
                    else:
                        print('Desconocido')
                        continue



                    dataResult = handleReport(values, reportId)
                    # Crear una consulta SQL para insertar datos
                    print(dataResult,caso)


                    if dataResult['vote'] in ['A', 'B', 'C', 'D', 'E', 'F'] and dataResult['control_id']<400:

                        sql = """UPDATE controls
                                SET `vote` = %s
                                WHERE `id` = %s"""
                        valores = (dataResult['vote'],dataResult['control_id'])
                        try:
                            # Ejecutar la consulta
                            cursor.execute(sql, valores)

                            # Guardar los cambios en la base de datos
                            conn.commit()
                        except Exception as e:
                            print(e)

                myFlag=1

    except KeyboardInterrupt:
        print("Interrupción recibida 1, deteniendo el script.")
        sendComandsToClose(device)


    except Exception as e:
        print("ERROR:" )
        print(e)
        return 1

    finally:
        try:
            os.remove(pid_file)
        except:
            pass



if __name__ == "__main__":
    pid_file = 'script_pid.txt'
    setProcessId()
    flag=1
    while flag:
        flag=main()
